Refactor some organization settings related code.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
This PR does some preparatory refactors for #21739.

- First commit just changes the variable name as discussed [here](https://github.com/zulip/zulip/pull/22770#discussion_r953181848).
- Second and third commits updates the id and name attributes of setting elements, so that we can remove extra code in further commits.
- Fourth commit changes `time_limit_dropdown_values` to a list to make it similar to the options list for notification batching period setting.
- Fifth commit is for extracting the code for showing and hiding custom input in a separate function.
- Sixth commit is for sending only updated settings in the API request data for message edit subsection.
- Seventh commit is to show save-discard widget only after custom input value is changed.
- Eigth commit is to extract code to set the msg edit and limit setting elements into a separate function.
- Last two commits are to refactor the code for realm-level default of email notification batching period setting such that we do not handle this setting as a special case.
<!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
Only user-facing changes are that the save-discard button is shown only after custom input is changed and also the flashing of save changes button mentioned in [this comment](https://github.com/zulip/zulip/pull/22863#issuecomment-1248348263) is fixed.
![msg-edit-refactor](https://user-images.githubusercontent.com/35494118/191534181-fe04248e-b605-46fb-b325-40ae2391dedd.gif)



**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

sahil839 commented:
I think it would take at least one more refactoring and then I can move forward with the remaining work of #21739. The refactoring I am talking about is for [this code](https://github.com/zulip/zulip/blob/378824204990b65cf5f0aa44dbd9ba39fec15c34/zerver/views/realm.py#L255) where we send all the three settings in event data and the api response irrespective of whether all the three settings are actually changed or not. `edit_topic_policy` setting would at least be independent when we implent the third bullet point of `Who can edit topics` heading in #21739, so we can definitely move it out. And since we have made the edit time limit setting and `allow_message_editing` setting independent in frontend, we can probably refactor this code to send only the changed setting in event data and api response.

timabbott commented:
That seems plausible regarding that API detail; can you start an api design conversation about it? We'll want to get the mobile/terminal folks input on that API change.

timabbott commented on static/js/settings_org.js:
```diff
@@ -341,12 +301,30 @@ function update_custom_value_input(property_name) {
 function set_msg_edit_limit_dropdown() {
-    const value = get_property_value("realm_message_content_edit_limit_seconds");
-    $("#id_realm_message_content_edit_limit_seconds").val(value.toString());
+    const dropdown_elem_val = get_time_limit_drodpown_setting_value(
```
`drodpown` => `dropdown`

sahil839 replied:
Fixed.

timabbott commented on static/js/settings_org.js:
```diff
@@ -813,18 +783,27 @@ function get_email_notification_batching_setting_element_value() {
 
     if (select_elem_val === "any_time") {
+        if (!for_api_data) {
+            return null;
```
Let's add a comment here.

sahil839 replied:
Added.

timabbott commented on static/js/settings_org.js:
```diff
@@ -813,18 +783,27 @@ function get_email_notification_batching_setting_element_value() {
+        $input_elem.parent().find(".admin-realm-time-limit-input"),
+    );
+    if (Number.isNaN(custom_input_val)) {
+        return null;
```
A comment here would also be helpful.

sahil839 replied:
Also did one more change here which I will elaborate in the comment below.

timabbott commented on static/templates/settings/organization_permissions_admin.hbs:
```diff
@@ -167,7 +167,7 @@
                         </label>
                         <input type="text" id="id_realm_message_content_edit_limit_minutes"
                           name="realm_message_content_edit_limit_minutes"
-                          class="admin-realm-time-limit-input prop-element"
```
Why is `prop-element` removed for these inputs? I forgot what exactly that does.

timabbott replied:
Oh, clarified in the commit message, makes sense!

timabbott commented:
I gave this a careful read and all the changes look correct to me; posted a few small comments on readability details. @alya can you review it as well?

sahil839 commented:
Added comments as asked above and also fixed the typo.

Regarding [this change](https://github.com/zulip/zulip/pull/23014#discussion_r977765597), previously when changing setting from `Anytime` to `Custom` the save-discard widget was hidden initially and then shown only when I type a valid input and was hidden when I type invalid input like alphabets. But since input value is actually changed, I changed it to show the save-discard widget also when the input is invalid and keep the save button disabled. The behavior after this change is shown below -
![custom-input-invalid](https://user-images.githubusercontent.com/35494118/191782495-528bf3ee-c74b-4f8f-b04f-683c94265946.gif)

alya commented:
Hm, how are we validating the custom fields? I don't think it's quite working right:

![Screen Shot 2022-09-22 at 12 18 38 PM](https://user-images.githubusercontent.com/2090066/191832882-de822281-bc9c-4fa6-bbbd-4796cf2e4efd.png)

Clicking "Save changes" in this state sets the time limit to 2 minutes.

alya commented:
Otherwise looks good to me!

sahil839 commented:
Added two commits to fix two things.
- First commit is to fix the bug mentioned by Alya above.
![custom-alpha-invalid](https://user-images.githubusercontent.com/35494118/191988472-e2793b2a-6042-4a12-b011-bfa8a2339258.gif)
- Second commit is to actually show the save-discard widget when input is invalid (like in the first case where it starts with number but is followed with alphabets) but with save button disabled.
![custom-invalid-save](https://user-images.githubusercontent.com/35494118/191988726-80a6c6b5-10e3-4dde-b1c3-110dc1958c5a.gif)

timabbott commented:
This is great, merged, thanks @sahil839!
